addons/account/models/partner.py

```diff
@@ -172,52 +172,34 @@ def write(self, vals):
-        null_country_dom = [('country_id', '=', False), ('country_group_id', '=', False)]
-
-        if zipcode:
-            zip_domain = [('zip_from', '<=', zipcode), ('zip_to', '>=', zipcode)]
-
-        if state_id:
-            state_domain = [('state_ids', '=', state_id)]
-
-        domain_country = base_domain + [('country_id', '=', country_id)]
-        domain_group = base_domain + [('country_group_id.country_ids', '=', country_id)]
-
-        # Build domain to search records with exact matching criteria
-        fpos = self.search(domain_country + state_domain + zip_domain, limit=1)
-        # return records that fit the most the criteria, and fallback on less specific fiscal positions if any can be found
-        if not fpos and state_id:
-            fpos = self.search(domain_country + null_state_dom + zip_domain, limit=1)
-        if not fpos and zipcode:
-            fpos = self.search(domain_country + state_domain + null_zip_dom, limit=1)
-        if not fpos and state_id and zipcode:
-            fpos = self.search(domain_country + null_state_dom + null_zip_dom, limit=1)
-
-        # fallback: country group with no state/zip range
-        if not fpos:
-            fpos = self.search(domain_group + null_state_dom + null_zip_dom, limit=1)
-
-        if not fpos:
-            # Fallback on catchall (no country, no group)
-            fpos = self.search(base_domain + null_country_dom, limit=1)
-        return fpos
-
     def _get_vat_valid(self, delivery, company=None):
         """ Hook for determining VAT validity with more complex VAT requirements """
         return bool(delivery.vat)
 
-    def _prepare_fpos_base_domain(self, vat_required):
+    def _get_fpos_ranking_functions(self, partner):
         return [
-            *self._check_company_domain(self.env.company),
-            ('auto_apply', '=', True),
-            ('vat_required', '=', vat_required),
+            ('vat_required', lambda fpos: (
+                not fpos.vat_required
+                or (self._get_vat_valid(partner, self.env.company) and 2)
+            )),
+            ('company_id', lambda fpos: len(fpos.company_id.parent_ids)),
+            ('zipcode', lambda fpos:(
+                not (fpos.zip_from and fpos.zip_to)
+                or (partner.zip and (fpos.zip_from <= partner.zip <= fpos.zip_to) and 2)
+            )),
+            ('state_id', lambda fpos: (
+                not fpos.state_ids
+                or (partner.state_id in fpos.state_ids and 2)
+            )),
+            ('country_id', lambda fpos: (
+                not fpos.country_id
+                or (partner.country_id == fpos.country_id and 2)
+            )),
+            ('country_group', lambda fpos: (
+                not fpos.country_group_id
+                or (partner.country_id in fpos.country_group_id.country_ids and 2)
+            )),
+            ('sequence', lambda fpos: fpos.sequence),
```

Comment:
Tests are probably failing because there is no default value, so we need it to not be falsy
```suggestion
            ('sequence', lambda fpos: fpos.sequence or 0.1),  # do not filter out sequence=0
```